### PR TITLE
Add Synonyms component

### LINKED ISSUE #622: Create synonym input field
We should create an input field for synonyms. For styling see https://github.com/Yoast/wordpress-seo-premium/issues/1796

### PR DESCRIPTION
**Pull request by nataliashitova**
## Summary

This PR can be summarized in the following changelog entry:

* Adds synonyms components.

## Relevant technical choices:

* Adds SynonymsSection and SynonymsInput components.
* Adds a synonym redux action and reducer.

## Test instructions

This PR can be tested by following these steps:

* Check the standalone under the tab "Keyword".
* Type something in the synonyms input and see the console log.

Fixes #622

### CODE REVIEW

File: app/KeywordExample.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+	 * @returns {ReactElement} The rendered keyword example.
+	 */
+	render() {
+		return (
+			<Container>
+				<KeywordInput
+					id="focus-keyword"
+					label={ "Focus keyword"}
+				/>
+
```
Comment (maartenleenders) [inline app/KeywordExample.js]: Does not require immediate fixing, but once the beta's have been made, this whitespace could be removed.

File: app/KeywordExample.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+	 * Renders an example of how to use the keyword inputs.
+	 *
+	 * @returns {ReactElement} The rendered keyword example.
+	 */
+	render() {
+		return (
+			<Container>
+				<KeywordInput
+					id="focus-keyword"
+					label={ "Focus keyword"}
```
Comment (maartenleenders) [inline app/KeywordExample.js]: Does not require immediate fixing, but once the beta's have been made, the inconsistent spacing around the label could be addressed.

File: composites/Plugin/Synonyms/actions/synonyms.js [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+export const insertSynonyms = function( index, synonyms ) {
+	return {
+		type: INSERT_SYNONYMS,
+		index,
+		synonyms,
+	};
+};
+
+/**
+ * An action creator for adding synonyms.
```
Comment (maartenleenders) [inline composites/Plugin/Synonyms/actions/synonyms.js]: Does not require immediate fixing, but once the beta's have been made, this description should be changed to "...for changing synonyms."

File: composites/Plugin/Synonyms/reducers/synonyms.js [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+function indexInArray( arr, index ) {
+	return index < 0 && index > arr.length;
+}
+
+/* eslint-disable complexity */
+/**
+ * A reducer for the synonyms.
+ *
+ * @param {Object} state  The current state of the object.
+ * @param {Object} action The current action received.
```
Comment (maartenleenders) [inline composites/Plugin/Synonyms/reducers/synonyms.js]: Does not require immediate fixing, but once the beta's have been made, this JSdoc could be improved for clarity by specifiying what the actions could contain. If possibly with redux, that is, haven't thought it through.

File: composites/Plugin/Synonyms/reducers/synonyms.js [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+/**
+ * Checks if an index is inside the array.
+ *
+ * @param {array}  arr   The array to check in.
+ * @param {number} index The index to check for.
+ *
+ * @returns {boolean} Whether the index exists in the array or not.
+ */
+function indexInArray( arr, index ) {
```
Comment (maartenleenders) [inline composites/Plugin/Synonyms/reducers/synonyms.js]: Looks like this is indexNotInArray.

Comment (atimmer) [reply composites/Plugin/Synonyms/reducers/synonyms.js]: I would avoid negations in function names.

File: composites/Plugin/Synonyms/reducers/synonyms.js [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+		case REMOVE_SYNONYMS:
+			if ( indexInArray( state, action.index ) ) {
+				return state;
+			}
+
+			return [
+				...state.slice( 0, action.index ),
+				...state.slice( action.index + 1 ),
+			];
```
Comment (maartenleenders) [inline composites/Plugin/Synonyms/reducers/synonyms.js]: In general, insert, change, and remove, can all be done by `Array.prototype.splice` https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/splice

File: package.json [hunk 7]
```diff
@@ -64,7 +64,7 @@
     "striptags": "^3.1.0",
     "styled-components": "^2.1.2",
     "whatwg-fetch": "^1.0.0",
-    "yoastseo": "^1.34.0"
+    "yoastseo": "https://github.com/Yoast/YoastSEO.js#stories/622-create-synonym-input-field"
```
Comment (maartenleenders) [inline package.json]: do not forget to change this back after creating the beta scripts

File: composites/Plugin/Synonyms/components/SynonymsSection.js [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+import { __ } from "@wordpress/i18n";
+
+import StyledSection from "../../../../forms/StyledSection/StyledSection";
+import SynonymsInput from "./SynonymsInput";
+import HelpText from "../../Shared/components/HelpText";
+import { makeOutboundLink } from "../../../../utils/makeOutboundLink";
+
+const HelpTextLink = makeOutboundLink();
+
+const Section = styled( StyledSection )`
```
Comment (maartenleenders) [inline composites/Plugin/Synonyms/components/SynonymsSection.js]: Does not require immediate fixing, but calling a styled StyledSection a normal Section is confusing.

File: composites/Plugin/Synonyms/components/SynonymsSection.js [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+
+class SynonymsSection extends React.Component {
+	/**
+	 * Constructs a SynonymsSection component
+	 *
+	 * @param {Object} props          The props for this input field component.
+	 * @param {string} props.id       The id of the SynonymsField.
+	 * @param {string} props.label    The label of the SynonymsField.
+	 * @param {string} props.synonyms The initial synonyms passed to the state.
+	 * @param {string} props.onChange The change event handler.
```
Comment (maartenleenders) [inline composites/Plugin/Synonyms/components/SynonymsSection.js]: probably a function?

File: composites/Plugin/Synonyms/components/SynonymsSection.js [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+			<Section
+				headingText={ label }
+				headingIcon={ "key" }
+				headingLevel={ 3 }
+				className={ "yoast-section" }
+			>
+				<HelpText
+					text={ helpText }
+				/>
+				<SynonymsInput
```
Comment (maartenleenders) [inline composites/Plugin/Synonyms/components/SynonymsSection.js]: Should investigate later: does this input field have a clickable label (as in forms that have a labelFor etc). It is expected functionality that when you click an input field's label it focuses the inputfield.

File: composites/Plugin/Synonyms/reducers/synonyms.js (hunk 5 above)
Comment (maartenleenders) [reply composites/Plugin/Synonyms/reducers/synonyms.js]: Oh definitely. I would align the boolean with the function title though. `indexInArray` outputs `true` if it's not in array, which is odd... So I'd return the opposite value.

Comment (maartenleenders) [reply composites/Plugin/Synonyms/reducers/synonyms.js]: I also know you know 😝

**Comment by maartenleenders** [comment]
CR 👍 because of the urge to merge ©️ .
Code could use some cleaning afterwards though.

File: composites/Plugin/Synonyms/reducers/synonyms.js (hunk 6 above)
Comment (igorschoester) [reply composites/Plugin/Synonyms/reducers/synonyms.js]: Splice changes the original, where slice returns a new array. In reducers you should never alter the original state. More info here: https://github.com/reduxjs/redux/blob/master/docs/recipes/reducers/ImmutableUpdatePatterns.md

Comment (maartenleenders) [reply composites/Plugin/Synonyms/reducers/synonyms.js]: I know about that. You could first load onto a new variable and splice it ofcourse :p.

But both ways work, just wanted to suggest a way to have less typing. Hence the ok.

File: composites/Plugin/Synonyms/components/SynonymsSection.js (hunk 10 above)
Comment (afercia) [reply composites/Plugin/Synonyms/components/SynonymsSection.js]: A bit late but yep, I'd recommend to implement this as done for the focus keyword input, with a `<label>` element associated with for/id attributes. In the plugin, the focus keyword input is then visually hidden with a `screen-reader-text` class. We can always decide later if the labels should be visually hidden or not.

File: composites/Plugin/Synonyms/components/SynonymsSection.js [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+	 * @param {string} props.id       The id of the SynonymsField.
+	 * @param {string} props.label    The label of the SynonymsField.
+	 * @param {string} props.synonyms The initial synonyms passed to the state.
+	 * @param {string} props.onChange The change event handler.
+	 *
+	 * @returns {void}
+	 */
+	constructor( props ) {
+		super( props );
+	}
```
Comment (afercia) [inline composites/Plugin/Synonyms/components/SynonymsSection.js]: Asked @atimmer for confirmation as this kind of usage of constructor seems unnecessary to me. And yes, the whole constructor could be removed in a further iteration.

**Pull request merged**